src/history/hs.c

```diff
@@ -529,7 +529,9 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
         ret = 0;
         goto done;
     }
-    while ((ret = __hs_delete_key_from_pos(session, cursor, btree->id, key)) == WT_RESTART)
+
+    while (
+      (ret = __hs_delete_key_from_pos(session, cursor, upd->txnid, btree->id, key)) == WT_RESTART)
```

Comment:
Actually, here should be first_non_ts_upd->txnid

Reply:
`first_non_ts_upd` isn't a variable at this point?

Reply:
We need to pass the value in here.

Reply:
Or we need to do some refactoring to move the remove code outside this function.

Reply:
This is starting to sound like something that should be handled in a different ticket, I'm not clear on exactly what the issue is here.

Reply:
The problem is we decide to remove the content in the history store once we see first_non_ts_upd but we do that when we insert a possibly older non-timestamped update. The easy solution is just to pass  `first_non_ts_upd ` into that function. A neater one I would recommend to do a refactor to move the deletion code outside this function.

Reply:
I'm happy to go for the easy one and have another ticket to refactor this.

Reply:
The easy one being the current implementation?

Reply:
yes

Reply:
No, the easy way is to modify the current implementation and pass first_non_ts_upd to __hs_insert_record_with_btree

Reply:
Where was it decided that we should use the transaction ID from the first_non_ts_upd? Should we use the original transaction ID from an update? If not, then I think that deserves a big comment somewhere, and/or a write up in the JIRA ticket.